// @cdkglobal/fortellis-dev-utils/src/middleware/trafficLogger.js
/*
 * ExpressJS Middleware for easily and consistently logging requests and responses across fortellis services
 * @param {function} logFn - function to call with the log object. Could be logger.info, console.log etc
 * @returns {function} Express middleware function
 */

function trafficLogger(logFn) {
  return (req, res, next) => {
    try {
      logFn({
        message: 'Request received',
        request: {
          method: req.method,
          url: req.originalUrl
        }
      });

      res.on('finish', () => {
        try {
          logFn({
            message: 'Response returned',
            response: {
              status: res.statusCode,
              message: res.statusMessage
            }
          });
        } catch (err) {
          console.log({
            timestamp: new Date().toISOString(),
            level: 'info',
            context: 'TRAFFIC_LOGGER_MIDDLEWARE',
            message: 'Response logging failed',
            error: err
          });
        }
      });
    } catch (err) {
      console.log({
        timestamp: new Date().toISOString(),
        level: 'info',
        context: 'TRAFFIC_LOGGER_MIDDLEWARE',
        message: 'Request logging failed',
        error: err
      });
    } finally {
      next();
    }
  };
}

module.exports = trafficLogger;
