// @stoplight/spectral/dist/functions/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const json_1 = require("@stoplight/json");
const AJV = require("ajv");
const jsonSpecv4 = require("ajv/lib/refs/json-schema-draft-04.json");
const jsonSpecv6 = require("ajv/lib/refs/json-schema-draft-06.json");
const jsonSpecv7 = require("ajv/lib/refs/json-schema-draft-07.json");
const lodash_1 = require("lodash");
const oasFormatValidator = require('ajv-oai/lib/format-validator');
const betterAjvErrors = require('better-ajv-errors/lib/modern');
const logger = {
    warn(...args) {
        const firstArg = args[0];
        if (typeof firstArg === 'string') {
            if (firstArg.startsWith('unknown format'))
                return;
            console.warn(...args);
        }
    },
    log: console.log,
    error: console.error,
};
const ajv = new AJV({
    meta: false,
    schemaId: 'auto',
    jsonPointers: true,
    unknownFormats: 'ignore',
    logger,
});
ajv.addMetaSchema(jsonSpecv4);
ajv.addMetaSchema(jsonSpecv6);
ajv.addMetaSchema(jsonSpecv7);
ajv._opts.defaultMeta = jsonSpecv4.id;
ajv._refs['http://json-schema.org/schema'] = 'http://json-schema.org/draft-04/schema';
ajv.addFormat('int32', { type: 'number', validate: oasFormatValidator.int32 });
ajv.addFormat('int64', { type: 'number', validate: oasFormatValidator.int64 });
ajv.addFormat('float', { type: 'number', validate: oasFormatValidator.float });
ajv.addFormat('double', { type: 'number', validate: oasFormatValidator.double });
ajv.addFormat('byte', { type: 'string', validate: oasFormatValidator.byte });
function getSchemaId(schemaObj) {
    if ('$id' in schemaObj) {
        return schemaObj.$id;
    }
    if ('id' in schemaObj) {
        return schemaObj.id;
    }
}
const validators = new class extends WeakMap {
    get(schemaObj) {
        const schemaId = getSchemaId(schemaObj);
        let validator = schemaId !== void 0 ? ajv.getSchema(schemaId) : void 0;
        if (validator !== void 0) {
            return validator;
        }
        validator = super.get(schemaObj);
        if (validator === void 0) {
            validator = ajv.compile(schemaObj);
            super.set(schemaObj, validator);
        }
        return validator;
    }
}();
const cleanAJVErrorMessage = (message, path, suggestion) => {
    const cleanMessage = typeof path === 'string' ? message.trim().replace(new RegExp(`^${lodash_1.escapeRegExp(path)}:\\s*`), '') : message.trim();
    return `${cleanMessage}${typeof suggestion === 'string' && suggestion.length > 0 ? `. ${suggestion}` : ''}`;
};
exports.schema = (targetVal, opts, paths) => {
    const results = [];
    const path = paths.target || paths.given;
    if (targetVal === void 0)
        return [
            {
                path,
                message: `${paths ? path.join('.') : 'property'} does not exist`,
            },
        ];
    const { schema: schemaObj } = opts;
    try {
        const validator = validators.get(schemaObj);
        if (!validator(targetVal) && validator.errors) {
            try {
                results.push(...betterAjvErrors(schemaObj, targetVal, validator.errors, { format: 'js' }).map(({ suggestion, error, path: errorPath }) => ({
                    message: cleanAJVErrorMessage(error, errorPath, suggestion),
                    path: [...path, ...(errorPath ? errorPath.replace(/^\//, '').split('/') : [])],
                })));
            }
            catch (_a) {
                results.push(...validator.errors.map(({ message, dataPath }) => ({
                    message: message ? cleanAJVErrorMessage(message, dataPath, void 0) : '',
                    path: [
                        ...path,
                        ...dataPath
                            .split('/')
                            .slice(1)
                            .map(json_1.decodePointerFragment),
                    ],
                })));
            }
        }
    }
    catch (ex) {
        if (ex instanceof AJV.MissingRefError) {
            results.push({
                message: ex.message,
                path,
            });
        }
        else {
            throw ex;
        }
    }
    return results;
};
